src/object/srv_obj.c

```diff
@@ -916,6 +925,77 @@ obj_log_csum_err(void)
 	bio_log_csum_err(bxc, info->dmi_tgt_id);
 }
 
+static void
+map_add_recx(daos_iom_t *map, const struct bio_iov *biov, uint64_t byte_idx)
+{
+	map->iom_recxs[map->iom_nr].rx_idx = byte_idx / map->iom_size;
+	map->iom_recxs[map->iom_nr].rx_nr = biov->bi_data_len / map->iom_size;
+	map->iom_nr++;
+}
+
+/** create maps for actually written to extents. */
+static int
+obj_fetch_create_maps(crt_rpc_t *rpc, struct bio_desc *biod)
+{
+	struct obj_rw_in	*orw = crt_req_get(rpc);
+	struct obj_rw_out	*orwo = crt_reply_get(rpc);
+	daos_iom_t		*maps;
+	daos_iom_t		*map;
+	struct bio_sglist	*bsgl;
+	daos_iod_t		*iod;
+	struct bio_iov		*biov;
+	uint32_t		 iods_nr;
+	uint32_t		 i, j;
+	uint64_t		 byte_idx;
+	uint32_t		 map_recx_nr;
+
+	/**
+	 * Allocate memory for the maps. There will be 1 per iod
+	 * Will be freed in obj_rw_reply
+	 */
+	iods_nr = orw->orw_iod_array.oia_iod_nr;
+	D_ALLOC_ARRAY(maps, iods_nr);
+	if (maps == NULL)
+		return -DER_NOMEM;
+	for (i = 0; i <  iods_nr; i++) {
+		bsgl = bio_iod_sgl(biod, i); /** 1 bsgl per iod */
+		iod = &orw->orw_iod_array.oia_iods[i];
+		map = &maps[i];
+		map_recx_nr = bsgl->bs_nr_out - bio_sgl_holes(bsgl);
+
+		/** will be freed in obj_rw_reply */
+		D_ALLOC_ARRAY(map->iom_recxs, map_recx_nr);
+		if (map->iom_recxs == NULL) {
+			D_FREE(maps);
+			return -DER_NOMEM;
+		}
+
+		map->iom_size = iod->iod_size;
+		map->iom_type = iod->iod_type;
+
+		if (map->iom_type != DAOS_IOD_ARRAY ||
+			bsgl->bs_nr_out == 0)
+			continue;
+
+		byte_idx = 0;
+		for (j = 0; j < bsgl->bs_nr_out; j++) {
+			biov = bio_sgl_iov(bsgl, j);
+			if (!bio_addr_is_hole(&biov->bi_addr))
+				map_add_recx(map, biov, byte_idx);
+
+			byte_idx += biov->bi_data_len;
+		}
+		D_ASSERT(map_recx_nr == map->iom_nr);
+		map->iom_recx_lo = map->iom_recxs[0];
+		map->iom_recx_hi = map->iom_recxs[map->iom_nr - 1];
```

Comment:
just to confirm with @mchaarawi that in daos_iom_t, is the iom_recx_lo/_hi are included in iom_recxs array? (or should move _lo/hi from iom_recxs array?)

Reply:
they should be included. The way i had this API was that the lo/hi are always returned, and the iom_recx array is optional if the user allocated it.